Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,229 @@
+```python
+# (Docs.python.org, 2018)
+class ObjectEncoder(json.JSONEncoder):
+  def default(self, object):
+    if isinstance(object, object_type):
+      # Return an already serializable type e.g.
+      return str(object_type)
+    return json.JSONEncoder.default(self,object)
+```
+
+**Loading:**
+Once again utilizing the same type of structure as the encoder so the symmetry between the systems allows developers to easily understand both of the systems, a 3 major part system should be used for abstraction and encapsulation, the file reading, the middle-man "Loader", and the decoding that is reproducing the objects and displaying them. Once again this allows testing and mocking of data to be easier and echoes MVP.
+
+Tackling reading the file should be relatively easy as reading it in should be handled by the JSON module that is built into python. It has a basic load all and load line, likely better if load line is utilized similarly to how json.dumps is used. That way the key-value pair object can be reconstructed in a similar way to how it was constructed before encoding it. Then the library allows decoding for both native types and custom types.
+
+To decode a normal type it would be a simple call to json.JSONDecoder which will utilise the following table for translating back to python objects:
+
+![JSON Decoding Table](./json-decoding-table.png) "*(Docs.python.org, 2018)*"
+
+With these recreated key-value pairs the whole point of this is to recreate the interfaces and display them to the user, this will be handled partially by the middle-man "Loader" which will take each object recreated by the ProjectReader and find it's relevant decoder. This will be achieved by asking the DecoderFactory for the correct Decoder this will be done by passing a string that represents the name loaded in from the file, this name should be present in one of the Decoder's tag list (Very similar to the encoder tag list and should be the same value actual implementation is undefined). When the DecoderFactory finds the tag it is looking for it will pass the Decoder back to the "Loader"
+
+Then the Loader will pass to the decoder the key-value pair object of state that is required to recreate the object and restore it's state. The best way to show this even though it will largely be left up to interface implementers to do this, is with an example:
+```python
+class exampleDecoder():
+  def decode(self, dict):
+    TestUI(dict["arg1"], dict["args2"])
+    TestUI.ui.spinBox.setValue(dict["spinBox"]) 
+    TestUI.ui.lineEdit.setValue(dict["lineEdit"]) 
+    TestUI.ui.dateEdit.setValue(dict["dateEdit"]) 
+    TestUI.ui.radioButton.setValue(dict["radioButton"]) 
+    TestUI.ui.checkBox.setValue(dict["checkBox"])
+    TestUI.ui.label_7.setVisible(dict["label_7Visible"])
+    TestUI.show()
```

Comment:
I would leave out `.show` and return the TestUI object to the caller to decide when best to do that.

Reply:
Agreed :+1: